File: c4_linkedlist/L160.cpp
//
// Created by LiHeng on 2020/4/19.
//
#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* ha = headA;
        ListNode* hb = headB;

        while(ha != hb){ // 如果不相交，二者各走一遍对方的路，最终会同时还走向对方的终点都是空也会退出循环
            ha = (ha == NULL? headB:ha->next);
            hb = (hb == NULL? headA:hb->next);
        }
        return ha;
    }
};

File: python-algorithm/Big_Company_Recent/bytedance/L160.py
# -*- coding: utf-8 -*-

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        l1, l2 = headA, headB
        while l1 != l2:
            if l1 == None:
                l1 = headB
            else:
                l1 = l1.next
            if l2 == None:
                l2 = headA
            else:
                l2 = l2.next
        return l1